pkg/kritis/gcbsigner/signer.go

```
@@ -0,0 +1,123 @@
+)
+
+type Signer struct {
+	config *Config
+	client metadata.Fetcher
+}
+
+type Config struct {
+	Secret   secrets.Fetcher
+	Validate buildpolicy.ValidateFunc
+}
+
+func New(client metadata.Fetcher, c *Config) Signer {
+	return Signer{
+		client: client,
+		config: c,
+	}
+}
+
+type ImageBuildInfo struct {
+	BuildID   string
+	ImageRef  string
+	BuiltFrom string
+}
+
+// For testing
+var (
+	authFetcher = authority.Authorities
+)
+
+// ValidateAndSign validates builtFrom against the build policies and creates
+// attestations for all authorities for the matching policies.
+// Returns an error if creating an attestation for any authority fails.
+func (r Signer) ValidateAndSign(buildInfo ImageBuildInfo, bps []v1beta1.BuildPolicy) error {
+	for _, bp := range bps {
+		glog.Infof("Validating %q against BuildPolicy %q", buildInfo.ImageRef, bp.Name)
+		if result := r.config.Validate(bp, buildInfo.BuiltFrom); result != nil {
+			glog.Errorf("Image %q does not match BuildPolicy %q: %s", buildInfo.ImageRef, bp.ObjectMeta.Name, result)
+			continue
+		}
+		glog.Infof("Image %q matches BuildPolicy %s, creating attestations", buildInfo.ImageRef, bp.Name)
+		if err := r.addAttestation(buildInfo.ImageRef, bp.Namespace, bp.Spec.AttestationAuthorityName); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r Signer) addAttestation(image string, ns string, authority string) error {
+	// Get all AttestationAuthorities in this namespace.
+	auths, err := authFetcher(ns)
+	if err != nil {
+		return err
+	}
+	if len(auths) == 0 {
+		return fmt.Errorf("no attestation authorities configured for namespace %s", ns)
+	}
+	errMsgs := []string{}
+	for _, a := range auths {
+		if a.ObjectMeta.Name == authority {
+			glog.Infof("Ceate attestation by %q for %q", image, authority)
+			// Get or Create Note for this this Authority
+			n, err := r.getOrCreateAttestationNote(&a)
+			if err != nil {
+				glog.Errorf("Error getting note: %s", err)
```

I'm generally not a fan of logging an error and returning it upward, because it typically results in duplicate messages in the error log. Consider whether this is behavior is useful in your case.

---

Yeah, leftover when I was tracing where the error was swallowed.  Removed.